src/date.rs

```diff
@@ -84,145 +83,108 @@ impl<Tz: TimeZone> Date<Tz> {
     /// Makes a new `DateTime` from the current date and given `NaiveTime`.
     /// The offset in the current date is preserved.
     ///
-    /// Panics on invalid datetime.
+    /// Returns `Err(Error)` on invalid datetime.
     #[inline]
-    pub fn and_time(&self, time: NaiveTime) -> Option<DateTime<Tz>> {
+    pub fn and_time(&self, time: NaiveTime) -> Result<DateTime<Tz>, Error> {
         let localdt = self.naive_local().and_time(time);
-        self.timezone().from_local_datetime(&localdt).single()
+        self.timezone().from_local_datetime(&localdt)?.single()
     }
 
     /// Makes a new `DateTime` from the current date, hour, minute and second.
     /// The offset in the current date is preserved.
     ///
-    /// Panics on invalid hour, minute and/or second.
-    #[deprecated(since = "0.4.23", note = "Use and_hms_opt() instead")]
+    /// Returns `Err(Error)` on invalid hour, minute and/or second.
     #[inline]
-    pub fn and_hms(&self, hour: u32, min: u32, sec: u32) -> DateTime<Tz> {
-        self.and_hms_opt(hour, min, sec).expect("invalid time")
-    }
-
-    /// Makes a new `DateTime` from the current date, hour, minute and second.
-    /// The offset in the current date is preserved.
-    ///
-    /// Returns `None` on invalid hour, minute and/or second.
-    #[inline]
-    pub fn and_hms_opt(&self, hour: u32, min: u32, sec: u32) -> Option<DateTime<Tz>> {
-        NaiveTime::from_hms_opt(hour, min, sec).and_then(|time| self.and_time(time))
+    pub fn and_hms(&self, hour: u32, min: u32, sec: u32) -> Result<DateTime<Tz>, Error> {
```

Comment:
IIUC, it looks like the recently added API `fn and_hms_opt` are being removed, and the "old" API like `fn and_hms` (previously marked `#[deprecated(...)]`) is being resurrected.

I'm afraid this will be too much churn for users. They will have switched from `fn and_hms` to `fn and_hms_opt`. With this PR, they must now switch back to `fn and_hms`.

My recommendation is to:

1. leave `fn and_hms_opt` in-place, so to not annoy current users of that API.
2. Then rewrite `fn and_hms` with the changes proposed here (returning `Result<DateTime<...>, Error>`).
3. The `fn and_hms_opt` should wrap a call to `fn and_hms`.

Of course, this should apply to all similar interfaces.

Reply:
Thanks for your input! I can not give specific reasoning for the case of `fn and_hms` and `fn and_hms_opt`, but in general I try to follow this pattern:

Return `Result<T,E>` for fallible functions and return `T` for "safe" functions, with "safe" being explicitly mentioned/required by the documentation or `const` context.

Prefixing all fallible functions with `try_` was also suggested in #815, but this seems excessive when applied consequently. Once the error-propagation is done, the backwards-compability may be more in focus. This targets v0.5 and will introduce breaking changes.

The mess with `fn and_hms` might also be a byproduct of the rebase, which I am still working on and might solve itself later in the progress. The functions related to `with_ymd_and_hms` also have a pending cleanup.
<hr>

Right now I'm stuck on `Timezone::offset_from_utc_date()` and `Timezone::offset_from_utc_datetime()`. Both are described as "cannot fail", although they should. Neither are deprecated.

`TimeZone::from_utc_date()` is deprecated and may panic/fail. `TimeZone::from_utc_datetime()` is suggested instead and may panic/fail.

It is hard to judge what behavior is intended and sometimes I have to guess or wait for feedback.

Reply:
> This targets v0.5 and will introduce breaking changes.

I guess it's up to @djc and @esheppa how many breaking changes they want in 0.5 release. My concern is overwhelming users with a seeming "back and forth" approach to `and_hms` and `and_hms_opt`. (_use `and_hms`! don't use `and_hms`, switch to `and_hms_opt`! use `and_hms` because `and_hms_opt` was removed!_)

> The mess with `fn and_hms` might also be a byproduct of the rebase

Sorry, I don't know what you mean.

> Right now I'm stuck on `Timezone::offset_from_utc_date()` and `Timezone::offset_from_utc_datetime()`. Both are described as "cannot fail", although they should. Neither are deprecated.

Good point. Seems like trying to be consistent about "cannot fail" and "should fail" might be a bit of work.

Reply:
> My concern is overwhelming users with a seeming "back and forth" approach to `and_hms` and `and_hms_opt`.

I picked up your proposal and added the  wrapper for backwards compatibility. This could be applied in other cases as well. However many function had a fingerprint change without receiving a new function name (returning `Result<T,E>` instead of just `T`). This might lead to additional confusion ...

```
    /// Wraps `and_hms` for backwards compability.
    #[deprecated(since = "0.5.0", note = "Use and_hms instead")]
    pub fn and_hms_opt(&self, hour: u32, min: u32, sec: u32) -> Result<DateTime<Tz>, Error> {
        self.and_hms(hour, min, sec)
    }
```

> > The mess with `fn and_hms` might also be a byproduct of the rebase
> 
> Sorry, I don't know what you mean.

The main trunk code base has diverged quite a lot since the original patch in #817. For many conflicts I chose the branch version and therefor dropped changes from the trunk. A few functions have been deprecated in the main trunk while they were still heavily used in the branch (e.g. `Utc.ymd()` ). I am not sure if `and_hms_opt` was present in the old branch already.

Reply:
> > My concern is overwhelming users with a seeming "back and forth" approach to `and_hms` and `and_hms_opt`.

> I picked up your proposal and added the wrapper for backwards compatibility. This could be applied in other cases as well. However many function had a fingerprint change without receiving a new function name (returning `Result<T,E>` instead of just `T`). This might lead to additional confusion ...
> 
> ```
> /// Wraps `and_hms` for backwards compability.
> #[deprecated(since = "0.5.0", note = "Use and_hms instead")]
> pub fn and_hms_opt(&self, hour: u32, min: u32, sec: u32) -> Result<DateTime<Tz>, Error> {
>     self.and_hms(hour, min, sec)
> }
> ```

Notice that `fn and_hms_opt` (and similar functions) return `Option<T>`.

https://github.com/chronotope/chrono/blob/v0.4.23/src/date.rs#L109-L111

Whereas in your proposed change, which you provided a sample of, it is returning `Result<T>`. That would still require users to change every call to `fn and_hms_opt` (and similar functions). So not very "backwards compatible". The `fn and_hms_opt` (and similar functions) should continue to return `Option<T>`.

Reply:
Whoopsie, changed in [33c0647](https://github.com/chronotope/chrono/pull/996/commits/33c06475d4b6a06ccf146b2f6ae3078320cb0066).